### ISSUE TITLE
Nested InitialTransitions calls superstate onEntry twice

### ISSUE THREAD

**Issue by ahorn42** [dotnet-state-machine/stateless#292]
Hi,

I use two nested InitialTransitions like in the TestCase `EntersSubStateofSubstate`

I assume the state machine of this test case looks like (BTW the generated dot file looks a lot different):
![stateless-test](https://user-images.githubusercontent.com/5106205/48953498-58b7f480-ef46-11e8-800d-5645ade105cc.png)

I modified the test case and added some OnEntry and OnExit actions to keep track of their execution:

```csharp

[Fact]
public void EntersSubStateofSubstate()
{
    var sm = new StateMachine<State, Trigger>(State.A);

    sm.Configure(State.A)
        .OnEntry(() => Debug.WriteLine("A OnEntry"))
        .OnExit(() => Debug.WriteLine("A OnExit"))
        .Permit(Trigger.X, State.B);

    sm.Configure(State.B)
        .OnEntry(() => Debug.WriteLine("B OnEntry"))
        .OnExit(() => Debug.WriteLine("B OnExit"))
        .InitialTransition(State.C);

    sm.Configure(State.C)
        .OnEntry(() => Debug.WriteLine("C OnEntry"))
        .OnExit(() => Debug.WriteLine("C OnExit"))
        .InitialTransition(State.D)
        .SubstateOf(State.B);

    sm.Configure(State.D)
        .OnEntry(() => Debug.WriteLine("D OnEntry"))
        .OnExit(() => Debug.WriteLine("D OnExit"))
        .SubstateOf(State.C);

    sm.Fire(Trigger.X);

    Assert.Equal(State.D, sm.State);
}
```

What I expect as output is:
```
A OnExit
B OnEntry
C OnEntry
D OnEntry
```

But what I get is:
```
A OnExit
B OnEntry
C OnEntry
B OnEntry
C OnEntry
D OnEntry
```

In my opinion this are one `B OnEntry` and one `C OnEntry` to much ;)

I browsed a little bit through the code but due to its complexity I didn't find something obvious till now what could cause this behavior.

Any ideas how to fix or work around this?

Best regards and thanks for doing stateless,
ahorn

**Comment by ahorn42** [comment]
After some debugging I found a rather simple solution to fix the problem, see PR #293.

I am not sure if something needs to be done with the 'async' methods, as I do not use nor understand them fully till now (according to stateless). But if so, I hope the solution is applicable there too.

All tests pass, so I hope it does not break anything :)

**Comment by HenningNT** [comment]
Sorry about not responding before, it's been busy…
Thank you for taking the time to look into this, this will be fixed in 4.4.
